### PR TITLE
AP_MotorsHeli: add thrust linearization for ddfp tail rotors

### PR DESCRIPTION
**Pull request by bnsgeyer**
This PR includes code from the multicopter motors library that linearizes the DDFP tail rotor thrust and adjusts for battery voltage.  This was tested on an actual heli.

### CODE REVIEW

**Comment by jasclarke308** [comment]
Could this be extended to supply voltage compensation for open loop throttle curve control of an electric main rotor? That is, when H_RSC_MODE=3?

**Comment by bnsgeyer** [comment]
@jasclarke308 I don't fly using the throttle curve with an electric heli.  Do you see the Rotor speed decrease for the same throttle request (i.e. collective position like hover) as the battery voltage decreases?  I suspect I could.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:747]: These could be moved up to the base class and shared.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:730]: Could be moved up to the base class and shared.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:625]: we should remove this commented out code..

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (lthall) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:742]: Maybe we should break this out as a function and use it in Multi too.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:729]: looks like the comment above this method should be updated.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:601]: probbily no harm in keeping this upto date even without having the tail type.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:771]: We log this on multirotor under "MOTB" you might want to add logging here too.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:147]: better as a subgroup to save indexes, possibly with DDFP_EXPO_EN ?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:742]: I guess ddfp_thrust_expo could be zero so we need to test that it is not zero.

**Comment by jasclarke308** [comment]
> @jasclarke308 I don't fly using the throttle curve with an electric heli. Do you see the Rotor speed decrease for the same throttle request (i.e. collective position like hover) as the battery voltage decreases? I suspect I could.

@bnsgeyer yes, with H_RSC_MODE=3 the rotor speed is proportional to voltage

**Comment by IamPete1** [comment]
@bnsgeyer @lthall This is what doing thrust linearization in the base class might look like. https://github.com/IamPete1/ardupilot/commit/39f48ae1bccbb4311107e76ac98c9394468f34d1 I have to admit it did not turn out as neatly as I had hoped. Maybe some shared math methods that take in all the arguments would be less intrusive but the battery filtering is a little tricky to do like that. Having said that this is not too bad and might be improved my moving the watt limit stuff into the new class as well. We could also tidy it up a bit adding helpers rather than reaching into the public variables of the new class. What do you think?

**Comment by bnsgeyer** [comment]
@IamPete1 I like what you've done.  I played with this a little last night but I think yours is the better way to go everything is contained in that class.  Mine was messier as I just moved the functions up to the parent class.  I didn't even get to the point of seeing if it would work.  You did miss some things like the declaration of the class in MotorsHeli_Single and the a call to update the lift_max.  I think it will need some work and testing but it looks like a good way to go.  Waiting to see what @rmackay9 and @lthall think.

**Comment by bnsgeyer** [comment]
@IamPete1 I think you should clean up what you have and PR it.  Please remove the heli stuff.  I will put that in after your PR goes in.  I think it would be easier that way since the changes to multi are a little more involved. @rmackay9  and @lthall what do you think?

**Comment by bnsgeyer** [comment]
This PR was rebased and modified to use the new methods supplied by PR #22787.  Requires testing before bringing it back to the DevCall

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:601]: No chance of a divide by zero here?

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:249]: better to grab these live from the SRV_Channel so users don't need to reboot on every change to min/max/trim

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by rmackay9** [comment]
In general this looks pretty good.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsHeli_Single.cpp:249]: Might as well just use set output scaled in that case.

Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli_Single.cpp:243]: @IamPete1  don’t think I can do that because I need to reference the input relative to the servo trim before sending it through linearization and compensation.

File: libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp
Comment (bnsgeyer) [inline libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp:58]: @IamPete1 i moved the parameter group into the thrust linearization class. Does that mess up the way you are adding these parameters for multi’s.

Comment (bnsgeyer) [reply libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp:69]: Everything looks good with heli parameters using this new table in the Thrust linearization class

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: added code to protect from divide by zero.  Thanks.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp
Comment (bnsgeyer) [reply libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: added param table to thrust linearization class and have heli using it as a subgroup.

**Comment by bnsgeyer** [comment]
@rmackay9 I think I am ready to merge this.  Wanted you to have a look before I click the merge button.  Please approve if you concur or let me know if you think I need to take it back to devcall.  I have tested this for heli in Realflight with a heli with DDFP.  I have not tested multirotor with the change with including a param table in the thrust linearization class.  But I figured the autotest would have caught that if there was a problem.

File: libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp
Comment (IamPete1) [inline libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp:78]: I think we need to define this out on multi copters and plane. Otherwise we will get the heli defaults from the param table rather than the copter ones.

Eventuality we can do the param conversion and go back to normal.

**Review by IamPete1 — CHANGES_REQUESTED** [review]

File: libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp
Comment (IamPete1) [reply libraries/AP_Motors/AP_Motors_Thrust_Linearization.cpp:93]: Sorry, no. 

Sorry for explaining badly, to elaborate on the problem slightly more. The issue is we load defaults to the same variables in the `AP_MotorsMulticopter` class here:

https://github.com/ArduPilot/ardupilot/blob/514434193d7c4dba259180db4a9aa83391c7cf01/libraries/AP_Motors/AP_MotorsMulticopter.cpp#L47-L75

And the rest further down. Then the same variables are initialised a second time when this object is created. So we have the variables with two defaults. This is because multirotor is not using the new param table but rather including the variables individually in the higher level table. 

So the fix is to either move to the new table on multirotor too, but then we need to do the param conversion. Or to remove the new table on multirotor. 

So you should be able to just do this:
```suggestion
#if APM_BUILD_TYPE(APM_BUILD_Heli)
    AP_Param::setup_object_defaults(this, var_info);
 #endif
```

You can check it worked by looking at the size change on copter, currently this adds 148 bytes. You can check in the build size action: 
https://github.com/ArduPilot/ardupilot/actions/runs/5242625088/jobs/9466260671?pr=22661

If copter is not using the table then the size change should be 0. 

Ultimately the longer term fix would be to move copter over to the new param table, I can look at adding that to this PR or as a future PR so we can remove this work around. There may also be a issue for quadplane with the depth of parameter trees that this avoids because heli is not supported on quadplane. Here its "copter -> motors -> thrust lin" on plane it would be "plane -> quadpalne -> motors -> thrust lin" that one level deeper might break params which would be one reason we could not use the new param table on copter with out a larger parameter tree depth fix.

**Review by IamPete1 — APPROVED** [review]
No size change on outside heli so I'm happy. Thanks!
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
ardusub          0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      399916
blimp            0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      662068
arducopter-heli  1116 (+0.0628%)  0 (0.0000%)  -4 (-0.0015%)  1116 (+0.0627%)                                  184256
arducopter       0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      191232
arduplane        0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      199632
antennatracker   0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      622184
ardurover        0 (0.0000%)      0 (0.0000%)  0 (0.0000%)    0 (0.0000%)                                      337072
```

**Pull request merged**
